### PR TITLE
chore: infer package manager when overriding with CDK

### PR DESCRIPTION
**Pull request by Lou1415926**
Will have to rebase on #5075

Resolve parts of #5050 on using yarn instead of npm for cdk override, with a small change from the [the first part of the comment I gave](https://github.com/aws/copilot-cli/issues/5050#issuecomment-1625851258).

I decided not include the drafted proposal of "reading package manager selection from env var"; instead, this PR just defaults to `npm` if after all the checks, there is still a tie. This is because I am not clear on whether it is a common design practice to use env vars to decide which package manager to use, and I expect that the current solution (check install + check project)  is already enough to resolve most of the case.

We can always add the enhancement of "reading package manager selection from env var" afterwards, if it's deemed needed.

This does not handle the second ask from the user, i.e. not assuming the location of `node_modules/`.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

**Comment by huanjani** [comment]
Related: #5050

File: internal/pkg/override/cdk.go [hunk 1]
```diff
@@ -172,6 +182,92 @@ func (cdk *CDK) cleanUp(in []byte) ([]byte, error) {
...
 }
 
+type packageManager struct {
+	name     string
+	lockFile string
+}
+
+var packageManagers = []packageManager{ // Alphabetically sorted based on name.
+	{
+		name:     "npm",
```
Comment (KollaAdithya) [inline internal/pkg/override/cdk.go:194]: can we use constants for `npm` and `yarn`
```
const (
    npm   = "npm"
    yarn  = "yarn"
)
```

File: internal/pkg/override/cdk.go [hunk 2]
```diff
@@ -172,6 +182,92 @@ func (cdk *CDK) cleanUp(in []byte) ([]byte, error) {
...
+		lockFile: "package-lock.json",
+	},
+	{
+		name:     "yarn",
+		lockFile: "yarn.lock",
+	},
+}
+
+func (cdk *CDK) installedPackageManager() ([]string, []error) {
```
Comment (KollaAdithya) [inline internal/pkg/override/cdk.go]: I am not completely sure if this works.

can we use `multierror.Append()` instead of returning `[]error`. If this does not work pls ignore this comment.

File: internal/pkg/override/cdk.go [hunk 3]
```diff
@@ -172,6 +182,92 @@ func (cdk *CDK) cleanUp(in []byte) ([]byte, error) {
...
+	}
+	var closestCandidate string
+	closestDistance := math.MaxInt
+	var errTargetNotFound *workspace.ErrTargetNotFound
+	for _, candidate := range packageManagers {
+		path, err := cdk.exec.Find(wd, 5, func(path string) (bool, error) {
+			return afero.Exists(cdk.fs, path)
+		}, candidate.lockFile)
+		if err == nil {
+			distance := strings.Count(wd, "/") - strings.Count(path, "/")
```
Comment (KollaAdithya) [inline internal/pkg/override/cdk.go]: Is it possible to use `filepath.Separator`
```
distance := strings.Count(wd, string(filepath.Separator)) - strings.Count(path, string(filepath.Separator))
```

File: internal/pkg/override/cdk.go (hunk 2 above)
Comment (Lou1415926) [reply internal/pkg/override/cdk.go]: good point, I addressed the feedback using `errors.Join`https://pkg.go.dev/errors#example-Join

File: internal/pkg/override/cdk.go (hunk 1 above)
Comment (Lou1415926) [reply internal/pkg/override/cdk.go:194]: I prefer keeping it as is if that's fine with you! Here is my reasoning -

I feel like we want to use `constant`s (instead of a "magic value") mostly in code logic:
```go
// Example: "npm" is a magic value
if _, err := cdk.exec.LookPath("npm"); err == nil {}

// Example: no magic value
npm := "npm"
if _, err := cdk.exec.LookPath(npm); err == nil {}
```

Using constant instead magic value makes it easier to refactor the code - you just need to change the value of `npm` to take care of all references to it. 

In this case, the `packageManagers` is a package-scoped variable instead of a part of "the logic". Consider this - if we use constants here, the code will looks like:
```go
const (
    npm   = "npm"
    yarn  = "yarn"
)

var packageManagers = []packageManager{
   {  name: npm, }
   {   name: yarn, }
}

if _, err := cdk.exec.LookPath(candidate.name); err == nil {}

```

The number of magic value did not decrease. All the logic code is still using `candidate.name` to refer to the package manager name. The only place that refers to `npm` is `packageManagers`.

Simply put, I feel like the problems with using "magic values" is already solved by `candidate.name` - and adding a layer of constant on top of `packageManagers` may not be a huge help.

Comment (Lou1415926) [reply internal/pkg/override/cdk.go:194]: But I did add this https://github.com/aws/copilot-cli/pull/5076/commits/7ad6de58ca07f4aa43640b4ea5f711e750e7093e that's to address your comment!

Comment (KollaAdithya) [reply internal/pkg/override/cdk.go:194]: This is really helpful. I got to learn when to use `const` and when not to

File: internal/pkg/override/cdk.go [hunk 4]
```diff
@@ -172,6 +181,95 @@ func (cdk *CDK) cleanUp(in []byte) ([]byte, error) {
...
+	for _, candidate := range packageManagers {
+		if _, err := cdk.exec.LookPath(candidate.name); err == nil {
+			installed = append(installed, candidate.name)
+		} else {
+			lookUpExecutableErrs = append(lookUpExecutableErrs, &errorExecutableNotFound{
+				executable: candidate.name,
+				error:      err,
+			})
+		}
+	}
```
Comment (dannyrandall) [inline internal/pkg/override/cdk.go]: seems like the only error case we care about is `len(installed) == 0` - should we remove this case and just return an error if we don't find _any_?

File: internal/pkg/override/cdk.go [hunk 5]
```diff
@@ -172,6 +181,95 @@ func (cdk *CDK) cleanUp(in []byte) ([]byte, error) {
...
+		name:     "npm",
+		lockFile: "package-lock.json",
+	},
+	{
+		name:     "yarn",
+		lockFile: "yarn.lock",
+	},
+}
+
+func (cdk *CDK) installedPackageManager() ([]string, error) {
```
Comment (dannyrandall) [inline internal/pkg/override/cdk.go]: nit
```suggestion
func (cdk *CDK) installedPackageManagers() ([]string, error) {
```

File: internal/pkg/override/cdk.go [hunk 6]
```diff
@@ -172,6 +181,95 @@ func (cdk *CDK) cleanUp(in []byte) ([]byte, error) {
...
+			lookUpExecutableErrs = append(lookUpExecutableErrs, &errorExecutableNotFound{
+				executable: candidate.name,
+				error:      err,
+			})
+		}
+	}
+	return installed, errors.Join(lookUpExecutableErrs...)
+}
+
+var getwd = os.Getwd
```
Comment (dannyrandall) [inline internal/pkg/override/cdk.go]: should this be a func on `CDK`?

File: internal/pkg/override/cdk.go [hunk 7]
```diff
@@ -172,6 +181,95 @@ func (cdk *CDK) cleanUp(in []byte) ([]byte, error) {
...
+	}
+	if len(installed) == 1 {
+		return installed[0], nil
+	}
+	manager, err := cdk.closestProjectManager()
+	if err != nil {
+		return "", err
+	}
+	if manager == "" {
+		manager = defaultPackageManager
```
Comment (dannyrandall) [inline internal/pkg/override/cdk.go]:
```suggestion
		return defaultPackageManager, nil
```

File: internal/pkg/override/errors.go [hunk 8]
```diff
@@ -7,17 +7,30 @@ import (
...
+}
+
+type errPackageManagerUnavailable struct {
+	parentError error
 }
 
-func (err *errNPMUnavailable) Error() string {
-	return fmt.Sprintf(`"npm" cannot be found: "npm" is required to override with the Cloud Development Kit: %v`, err.parent)
+func (err *errPackageManagerUnavailable) Error() string {
+	return fmt.Sprintf("cannot find a package manager to override with the Cloud Development Kit:\n%s", err.parentError.Error())
```
Comment (dannyrandall) [inline internal/pkg/override/errors.go]: nit - i think this is the case? does it only work with javascript package managers?
```suggestion
	return fmt.Sprintf("cannot find a javascript package manager to override with the Cloud Development Kit:\n%s", err.parentError.Error())
```

**Review by dannyrandall — COMMENTED** [review]
looks mostly good, just some small feedback!

File: internal/pkg/override/cdk.go (hunk 4 above)
Comment (Lou1415926) [reply internal/pkg/override/cdk.go]: Good point, I addressed this by doing the following instead:
```go
func (cdk *CDK) installedPackageManagers() ([]string, error) {
	var installed []string
	for _, candidate := range packageManagers {
		if _, err := cdk.exec.LookPath(candidate.name); err == nil {
			installed = append(installed, candidate.name)
		} else if !errors.Is(err, exec.ErrNotFound) { // Return unknown error! Swallow ErrNotFound.
			return nil, err
		}
	}
	return installed, nil
}


func (cdk *CDK) packageManager() (string, error) {
	installed, err := cdk.installedPackageManagers()
	if err != nil { // Handle unknown error.
		return "", err
	}
	if len(installed) == 0 {
		return "", &errPackageManagerUnavailable{}
	}
```

**Review by dannyrandall — APPROVED** [review]

**Review by KollaAdithya — APPROVED** [review]
LGTM!
